### ISSUE TITLE
Operator Does Not Respect Collector Memory Configuration

### ISSUE THREAD

**Issue by HaddadJoe** [GoogleCloudPlatform/prometheus-engine#235]
Hi everyone, thanks for making that project happen!

We are running the latest version of the managed Prometheus engine(v0.4.1) in a gke cluster version below using kubectl
```
client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.3", GitCommit:"816c97ab8cff8a1c72eccca1026f7820e93e0d25", GitTreeState:"clean", BuildDate:"2022-01-25T21:17:57Z", GoVersion:"go1.17.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.8-gke.200", GitCommit:"87bd5277f85536686030dd1fd7d6e39776dd44d1", GitTreeState:"clean", BuildDate:"2022-03-17T20:42:25Z", GoVersion:"go1.16.14b7", Compiler:"gc", Platform:"linux/amd64"}
```

Our collector is running out of memory (OOMKilled). After doing some digging, we found that there's some variables to control the memory of the collector. We tried bumping these by editing the operator's deployment but with no luck. In comparison, we tried to change the same variable for the evaluator, it worked just fine, the memory increase was reflected and the operator rolledout a new evaluator with the updated information. This never happened on the collector, even after manually deleting the daemonset to force a clean creation.

Here are the operator's arguments
```
  50       - args:
  51         - --public-namespace=gmp-public
  52         - --priority-class=gmp-critical
  53         - --image-collector=gke.gcr.io/prometheus-engine/prometheus:v2.28.1-gmp.7-gke.0
  54         - --image-config-reloader=gke.gcr.io/prometheus-engine/config-reloader:v0.4.1-gke.0
  55         - --image-rule-evaluator=gke.gcr.io/prometheus-engine/rule-evaluator:v0.4.1-gke.0
  56         - --webhook-addr=:10250
  57         - --collector-memory-limit=6000
  58         - --collector-memory-resource=3000
  50         - --evaluator-memory-resource=3000
```

1) We think a better documentation of the operator's configuration would be useful for other people facing the same problem. Mainly variables available [here](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/v0.4.1/cmd/operator/main.go#L85). Happy to open a PR if this is needed.
2) Not 100% sure if this is the problem but it looks like the collector's memory is hardcoded [here](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/v0.4.1/pkg/operator/collection.go#L350
) where as the evaluator's memory is being taken from the resource opts [here](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/v0.4.1/pkg/operator/operator_config.go#L428)
3) Any idea in the first place why our collector is consuming more than 3Gb? Is this expected? We only have 5 podmonitoring resource deployed with a scrape interval of 30 seconds. 

Thanks in advance for the help, happy to provide more debugging information.

**Comment by pintohutch** [comment]
Hi,

> 1. We think a better documentation of the operator's configuration would be useful for other people facing the same problem. Mainly variables available [here](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/v0.4.1/cmd/operator/main.go#L85). Happy to open a PR if this is needed.

Please do!

> 2. Not 100% sure if this is the problem but it looks like the collector's memory is hardcoded [here](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/v0.4.1/pkg/operator/collection.go#L350) w[here](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/v0.4.1/pkg/operator/operator_config.go#L428) as the evaluator's memory is being taken from the resource opts here

Thanks for reporting! This is indeed a bug as you noted in (2).

We're actually in the [process](https://github.com/GoogleCloudPlatform/prometheus-engine/pull/227) of removing these flags, favoring a yaml manifest where users can configure these things themselves without having to funnel everything through the operator.

In the meantime, if you're interested in opening a PR with the fix, that'd be helpful. We may cut a release before https://github.com/GoogleCloudPlatform/prometheus-engine/pull/227 is merged.

> 3. Any idea in the first place why our collector is consuming more than 3Gb? Is this expected? We only have 5 podmonitoring resource deployed with a scrape interval of 30 seconds.

The collector is a thinly [forked Prometheus](https://github.com/GoogleCloudPlatform/prometheus) that pushes data to Google Cloud Monitoring. Because of this we [document](https://cloud.google.com/stackdriver/docs/managed-prometheus#gmp-data-collection)

> Streaming data to Managed Service for Prometheus consumes additional resources. If you are manually deploying collectors, then we recommend increasing CPU and memory limits by 5x and adjusting them based on actual usage.

This is mainly attributed to buffering and serializing time series data protobuf messages before sending them over the wire. Some figures of interest would be knowing the time series volume being ingested by your OOMing collectors. For example, querying a collector via
```
% kubectl port-forward <ooming-collector-pod> -ngmp-system 19090

# Then access the Prometheus UI in a browser via http://localhost:19090 and query
prometheus_tsdb_head_series

and

rate(prometheus_tsdb_head_samples_appended_total[5m])
```
would give you some idea of volume load. Note: the collector is only scraping targets co-located on the same node - all metrics it's gathering are constrained to workloads on that node.

Two things you can try:
1. Increase scrape duration to > 30s, say 60s.
1. The OperatorConfig has a [`.Collection.Filter`](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/f8a719bba5ae91321ae4a783c5954bc94675f593/doc/api.md#collectionspec) field where you can filter out metrics from being cached and sent to Cloud Monitoring. Adding filters to discard unused metrics can help here (and can substantially lower your bill!)

Hope that helps

**Comment by raphapassini** [comment]
Hello @pintohutch I opened a [small PR](https://github.com/GoogleCloudPlatform/prometheus-engine/pull/236) fixing this bug. 

Thanks to everyone who made this project happened!

**Comment by raphapassini** [comment]
Hey @pintohutch thanks for the review.

There's any change we can have a new release any time soon?

**Comment by pintohutch** [comment]
Hi @raphapassini,

Apologies for the delayed response here. Yes, we will cut a release with this fix soon. I'll post here with an update when we're ready.

**Comment by HaddadJoe** [comment]
@pintohutch any updates here? Our collectors keep crashing due to OOM and with the inability to increase the memory (it has to be done through the operator) we're missing metrics and therefore affecting the "perceived SLA" of GMP.

**Comment by pintohutch** [comment]
Hi @HaddadJoe,

Yes! We actually have the latest release publicly available at (v0.4.2-gke.1)[https://github.com/GoogleCloudPlatform/prometheus-engine/releases/tag/v0.4.2-gke.1]. We're in the process of drafting release notes, but the manifest and images should be good to use now.

Wanna give them a try?

**Comment by lyanco** [comment]
This has been released publicly with the launch of v0.4.3-gke.0, which is rolling out now.
